src/puya/ir/arc4_router.py

```diff
@@ -70,6 +71,123 @@ def btoi(bytes_arg: awst_nodes.Expression, location: SourceLocation) -> awst_nod
+                op_code="substring",
+                immediates=[6, 8],
+                wtype=wtypes.bytes_wtype,
+                stack_args=[
+                    awst_nodes.IntrinsicCall(
+                        source_location=location,
+                        op_code="itob",
+                        stack_args=[
+                            awst_nodes.IntrinsicCall.bytes_len(base_temp.define, location)
+                        ],
+                        wtype=wtypes.bytes_wtype,
+                    )
+                ],
+            )
+            return awst_nodes.ReinterpretCast(
+                source_location=location,
+                wtype=wtypes.ARC4DynamicArray.from_element_type(
+                    wtypes.ARC4UIntN.from_scale(8, alias="byte")
+                ),
+                expr=awst_nodes.IntrinsicCall(
+                    source_location=location,
+                    op_code="concat",
+                    stack_args=[length, base_temp.read],
+                    wtype=wtypes.bytes_wtype,
+                ),
+            )
+        case wtypes.WTuple(types=types):
+            base_temp = create_temporary_assignment(base, location)
+
+            return awst_nodes.ARC4Encode(
+                source_location=location,
+                value=awst_nodes.TupleExpression.from_items(
+                    items=[
+                        arc4_encode(
+                            awst_nodes.TupleItemExpression(
+                                base=base_temp.define if i == 0 else base_temp.read,
+                                index=i,
+                                source_location=location,
+                            ),
+                            wtypes.avm_to_arc4_equivalent_type(t),
+                            location,
+                        )
+                        for i, t in enumerate(types)
+                    ],
+                    location=location,
+                ),
+                wtype=target_wtype,
+            )
+
+        case _:
+            return awst_nodes.ARC4Encode(
+                source_location=location,
+                value=base,
+                wtype=target_wtype,
+            )
+
+
+def arc4_decode(
+    bytes_arg: awst_nodes.Expression, target_wtype: wtypes.WType, location: SourceLocation
+) -> awst_nodes.Expression:
+    match bytes_arg.wtype:
+        case wtypes.ARC4DynamicArray(
+            element_type=wtypes.ARC4UIntN(n=8)
+        ) if target_wtype == wtypes.bytes_wtype:
+            return awst_nodes.IntrinsicCall(
+                wtype=wtypes.bytes_wtype,
+                op_code="extract",
+                immediates=[2, 0],
+                source_location=location,
+                stack_args=[
+                    awst_nodes.ReinterpretCast(
+                        expr=bytes_arg, wtype=wtypes.bytes_wtype, source_location=location
+                    )
+                ],
+            )
+        case wtypes.ARC4Tuple(types=tuple_types):
```

Comment:
Can we also handle nested tuples here easily? If not should probably disallow it in `has_arc4_equivalent_type`